### PR TITLE
Implement Room Privacy Modes

### PR DESCRIPTION
**Pull request by cblgh**
This PR tracks the progress of implementing support for the SSB Rooms 2 concept of [Privacy Modes](https://ssb-ngi-pointer.github.io/rooms2/#privacy-modes).

The privacy modes control which types of users are allowed to perform certain actions, such as creating invites. In the open mode, any person may visit the room and create an invite to it. In the restricted mode, only existing room members with the roles of admin or moderator may invite new users.

Privacy modes:
* Open
* Community
* Restricted

Closes https://github.com/ssb-ngi-pointer/go-ssb-room/issues/96

Outstanding todos:
* [ ] Create an open invite view (on the basic description page `index.html` if mode is open?)
* [x] Fix tests to take / create mock config (after sqlite database?)

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -81,6 +80,20 @@ func checkAndLog(err error) {
 	}
 }
 
+type Config struct {
+	// open, community, restricted
+	privacyMode roomdb.PrivacyMode
+}
+
+// cblgh: i dont get why we need a context here, if it is potentially a bad code smell?
```
Comment (cryptix) [inline cmd/server/main.go:88]: We can ignore it here (you can use `_ context.Context`) but the database code later will need it and changing interfaces later is just hassle no-one needs.

Comment (cblgh) [reply cmd/server/main.go:88]: ah yeah, i was kind of guessing at that (but i have no knowledge yet about the database stuff—i guess i will have that though when this is merged! exciting~~~)

File: muxrpc/handlers/tunnel/server/state.go [hunk 2]
```diff
@@ -70,6 +74,23 @@ func (h *Handler) endpoints(_ context.Context, req *muxrpc.Request, snk *muxrpc.
 		return err
 	}
 
+	// cblgh:
+	// * reject if key in deny list / DeniedKeysService
+	pm, err := h.config.GetPrivacyMode(nil)
```
Comment (cryptix) [inline muxrpc/handlers/tunnel/server/state.go:79]:
```suggestion
	pm, err := h.config.GetPrivacyMode(ctx)
```

File: muxrpc/handlers/tunnel/server/state.go [hunk 3]
```diff
@@ -70,6 +74,23 @@ func (h *Handler) endpoints(_ context.Context, req *muxrpc.Request, snk *muxrpc.
...
+		return fmt.Errorf("running with unknown privacy mode")
+	}
+
+	switch pm {
+	case roomdb.ModeCommunity:
+		fallthrough
+	case roomdb.ModeRestricted:
+		_, err := h.members.GetByFeed(ctx, *ref)
+		if err != nil {
+			return fmt.Errorf("external user tried to join room despite its elevated privacy mode")
```
Comment (cryptix) [inline muxrpc/handlers/tunnel/server/state.go:90]:
```suggestion
			return fmt.Errorf("external user are not allowed to enumerate members")
```

File: roomsrv/init_network.go [hunk 4]
```diff
@@ -27,7 +28,12 @@ func (s *Server) initNetwork() error {
 			return &s.master, nil
 		}
 
-		if _, err := s.authorizer.GetByFeed(s.rootCtx, *remote); err == nil {
+		pm, err := s.Config.GetPrivacyMode(nil)
+		if err != nil {
+			return nil, fmt.Errorf("running with unknown privacy mode")
+		}
+		// if privacy mode is open, allow all connections. otherwise, only allow members
+		if _, err := s.authorizer.GetByFeed(s.rootCtx, *remote); err == nil || pm != roomdb.ModeOpen {
```
Comment (cryptix) [inline roomsrv/init_network.go:36]: I'm not 100% sure this is correct, only in restricted mode it should be restricted to members. cc @staltz 

In community mode, everyone can connect (and use aliases to connect to people directly) but just members can see endpoints.

I'd do 

```go
if pm == roomdb.ModeRestricted {
  if _, err := s.authorizer.GetByFeed(s.rootCtx, *remote); err!=nil {
    return nil, fmt.Errorf("access restricted to members")  
  }
}
```


( fyi: `s.authorizer == members list` )

File: muxrpc/handlers/tunnel/server/state.go [hunk 5]
```diff
@@ -70,6 +74,23 @@ func (h *Handler) endpoints(_ context.Context, req *muxrpc.Request, snk *muxrpc.
 		return err
 	}
 
+	// cblgh:
+	// * reject if key in deny list / DeniedKeysService
```
Comment (cryptix) [inline muxrpc/handlers/tunnel/server/state.go:78]: the `DeniedKeysService` lookup needs to be added to `roomsrv/init_network.go`, not here.

I agree that endpoints is a confusion name, it's really just _potentiall endpoints_ and tangentiall to the actual connection building.

**Review by cryptix — CHANGES_REQUESTED** [review]
The idea and extension of struct's is correct..

Just some semantics of how it's supposed to work.

File: roomsrv/init_network.go (hunk 4 above)
Comment (staltz) [reply roomsrv/init_network.go:36]: > In community mode, everyone can connect (and use aliases to connect to people directly) but just members can see endpoints.

correct

File: muxrpc/handlers/tunnel/server/state.go (hunk 5 above)
Comment (staltz) [reply muxrpc/handlers/tunnel/server/state.go:78]: https://github.com/ssb-ngi-pointer/go-ssb-room/issues/89 :cat2:

File: roomsrv/init_network.go (hunk 4 above)
Comment (cblgh) [reply roomsrv/init_network.go:36]: ah alright, gotcha :)

File: web/handlers/admin/handler.go [hunk 6]
```diff
@@ -44,6 +44,7 @@ var HTMLTemplates = []string{
 // Databases is an option struct that encapsualtes the required database services
 type Databases struct {
 	Aliases       roomdb.AliasesService
+	Config        roomdb.RoomConfig // cblgh: kind of confusing that we have two identically named structs, in different http handler contexts?
```
[the remark this answers is not in the document]
Comment (cblgh) [reply web/handlers/admin/handler.go:47]: case of me thinking out loud, will just mark this as resolved :)

File: muxrpc/handlers/tunnel/server/state.go (hunk 5 above)
Comment (cblgh) [reply muxrpc/handlers/tunnel/server/state.go:78]: yeah i know, this is just a note to use it + remember what the service was called

File: roomsrv/init_network.go (hunk 4 above)
Comment (cblgh) [reply roomsrv/init_network.go]: 97cef32

File: roomdb/sqlite/config.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+	"github.com/ssb-ngi-pointer/go-ssb-room/roomdb"
+	"github.com/ssb-ngi-pointer/go-ssb-room/roomdb/sqlite/models"
+	"github.com/volatiletech/sqlboiler/v4/boil"
+)
+
+// cblgh: ask cryptix about the details of the syntax of the "compiler assertion" below
+// why two parens? this does not look like a typical type assertion? e.g. <var>.(type)
+// hm-maybe this is a type conversion, forcing "nil" to be a *Aliases?
+var _ roomdb.AliasesService = (*Aliases)(nil)
```
Comment (cryptix) [inline roomdb/sqlite/config.go:18]:
```suggestions
var _ roomdb.RoomConfig = (*Config)(nil)
```

File: roomdb/sqlite/config.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	// make sure the privacy mode is an ok value
+	err := pm.IsValid()
+	if err != nil {
+		return err
+	}
+
+	// cblgh: a walkthrough of this step (again, now that i have some actual context) would be real good :)
+	err = transact(c.db, func(tx *sql.Tx) error {
+		// get the settings row
+		config, err := models.FindConfig(ctx, c.db, configRowID)
```
Comment (cryptix) [inline roomdb/sqlite/config.go:59]:
```suggestion
		config, err := models.FindConfig(ctx, tx, configRowID)
```

**Review by cryptix — CHANGES_REQUESTED** [review]
Just two small things.

**Comment by staltz** [comment]
@cblgh I also linked issue #12 with this PR

**Pull request merged**